src/treq/auth.py

```diff
@@ -1,13 +1,223 @@
 # Copyright 2012-2020 The treq Authors.
 # See LICENSE for details.
 from __future__ import absolute_import, division, print_function
+import re
+import time
+import hashlib
 
 import binascii
-from typing import Union
+from typing import Union, Optional
+from urllib.parse import urlparse
 
+from twisted.python.randbytes import secureRandom
 from twisted.web.http_headers import Headers
-from twisted.web.iweb import IAgent
+from twisted.web.iweb import IAgent, IBodyProducer, IResponse
 from zope.interface import implementer
+from requests.utils import parse_dict_header
+
+
+_DIGEST_HEADER_PREFIX_REGEXP = re.compile(b'digest ', flags=re.IGNORECASE)
+
+
+def _generate_client_nonce(server_side_nonce: bytes) -> str:
+    return hashlib.sha1(
+        hashlib.sha1(server_side_nonce).digest() +
+        secureRandom(16) +
+        time.ctime().encode('utf-8')
+    ).hexdigest()[:16]
+
+
+def _md5_utf_digest(x: bytes) -> str:
+    return hashlib.md5(x).hexdigest()
+
+
+def _sha1_utf_digest(x: bytes) -> str:
+    return hashlib.sha1(x).hexdigest()
+
+
+def _sha256_utf_digest(x: bytes) -> str:
+    return hashlib.sha256(x).hexdigest()
+
+
+def _sha512_utf_digest(x: bytes) -> str:
+    return hashlib.sha512(x).hexdigest()
+
+
+class HTTPDigestAuth(object):
+    """
+    The container for HTTP Digest authentication credentials.
+
+    This container will cache digest auth parameters,
+    in order not to recompute these for each request.
+    """
+
+    def __init__(self, username: Union[str, bytes],
+                 password: Union[str, bytes]):
+        if not isinstance(username, bytes):
+            self._username = username.encode('utf-8')
+        else:
+            self._username = username
+        if not isinstance(password, bytes):
+            self._password = password.encode('utf-8')
+        else:
+            self._password = password
+
+        # (method,uri) --> digest auth cache
+        self._digest_auth_cache = {}
+
+    def build_authentication_header(
+            self, path: bytes, method: bytes, cached: bool, nonce: bytes,
+            realm: bytes, qop: Optional[Union[str, bytes]] = None,
+            algorithm: bytes = b'MD5', opaque: Optional[bytes] = None
+            ) -> bytes:
+        """
+        Build the authorization header for credentials got from the server.
+        Algorithm is accurately ported from https://github.com/psf/requests
+            with small adjustments.
+        See
+        https://github.com/psf/requests/blob/v2.5.1/requests/auth.py#L72
+            for details.
+
+        :param algorithm: algorithm to be used for authentication,
+            defaults to MD5, supported values are
+            "MD5", "MD5-SESS" and "SHA"
```

Comment:
Yikes. Is it even secure in this day and age to have an MD5-based authentication mechanism? I would have assumed digest auth would have moved on to some better hash algorithms. Are we going to get kicked out of FIPS-requiring applications because this string appears here?

I guess [the wikipedia page](https://en.wikipedia.org/wiki/Digest_access_authentication#Impact_of_MD5_security_on_digest_authentication) says it's still OK, but it also notes that in 2015, "SHA-256", "SHA-256-sess", "SHA-512-256" and "SHA-512-256-sess" were standardized.  Surely it would make more sense to default to one of those?  Or do we even need a default at all?

Also: this should really be an enum, not `bytes`, so that you won't get `UnknownDigestAuthAlgorithm` due to a typo.

Reply:
> Yikes. Is it even secure in this day and age to have an MD5-based authentication mechanism?

Apparently it's not that bad due to how it's constructed. Better than basic auth at least for systems not using https.

> I would have assumed digest auth would have moved on to some better hash algorithms. Are we going to get kicked out of FIPS-requiring applications because this string appears here?

In general I don't think there's been much adoption of newer digest auth features/algos. Python's built in hashlib provides MD5 support so if there's FIPS issues with MD5 this isn't going to make much difference there.

> I guess [the wikipedia page](https://en.wikipedia.org/wiki/Digest_access_authentication#Impact_of_MD5_security_on_digest_authentication) says it's still OK, but it also notes that in 2015, "SHA-256", "SHA-256-sess", "SHA-512-256" and "SHA-512-256-sess" were standardized. Surely it would make more sense to default to one of those? Or do we even need a default at all?

AFAIU the default must be MD5 for compatibility purposes for cases where the server doesn't explicitely define an algo. The algo choice is effectively entirely server controlled, the default is only for handling the non-explicitely defined case(which always means MD5 is required AFAIU).

> Also: this should really be an enum, not `bytes`, so that you won't get `UnknownDigestAuthAlgorithm` due to a typo.

Any specific implementation I should try and make the style similar to here?

Reply:
> Also: this should really be an enum, not `bytes`, so that you won't get `UnknownDigestAuthAlgorithm` due to a typo.

Changed to using an enum instead of `bytes`.